sw/airborne/subsystems/ins/ins_ekf2.cpp

```diff
@@ -0,0 +1,509 @@
+      stateSetPositionNed_f(&pos);
+
+      /* Get the velocity in NED frame */
+      float vel_f[3] = {};
+      struct NedCoor_f speed;
+      ekf.get_velocity(vel_f);
+      speed.x = vel_f[0];
+      speed.y = vel_f[1];
+      speed.z = vel_f[2];
+
+      // Publish to state
+      stateSetSpeedNed_f(&speed);
+
+      /* Get the accelrations in NED frame */
+      float vel_deriv_f[3] = {};
+      struct NedCoor_f accel;
+      ekf.get_vel_deriv_ned(vel_deriv_f);
+      accel.x = vel_deriv_f[0];
+      accel.y = vel_deriv_f[1];
+      accel.z = vel_deriv_f[2];
+
+      // Publish to state
+      stateSetAccelNed_f(&accel);
+
+      /* Get local origin */
+      // Position of local NED origin in GPS / WGS84 frame
+      struct map_projection_reference_s ekf_origin = {};
+      float ref_alt;
+      struct LlaCoor_i lla_ref;
+      uint64_t origin_time;
+
+      // Only update the origin when the state estimator has updated the origin
+      bool ekf_origin_valid = ekf.get_ekf_origin(&origin_time, &ekf_origin, &ref_alt);
+      if(ekf_origin_valid && (origin_time > ekf2.ltp_stamp)) {
+        lla_ref.lat = ekf_origin.lat_rad * 180.0 / M_PI *1e7; // Reference point latitude in degrees
+        lla_ref.lon = ekf_origin.lon_rad * 180.0 / M_PI *1e7; // Reference point longitude in degrees
+        lla_ref.alt = ref_alt * 1000.0;
+        ltp_def_from_lla_i(&ekf2.ltp_def, &lla_ref);
+        stateSetLocalOrigin_i(&ekf2.ltp_def);
+        
+        ekf2.ltp_stamp = origin_time;
+      }
+    }
+  }
+
+#if defined SITL && USE_NPS
+  if (nps_bypass_ins) {
+    sim_overwrite_ins();
+  }
+#endif
+
+  ekf2.got_imu_data = false;
+}
+
+/** Publish the attitude and get the new state
+ * Directly called after a succeslfull gyro+accel reading
+ */
+static void ins_ekf2_publish_attitude(uint32_t stamp) {
+  float rates_int[3], accel_int[3];
+  rates_int[0] = ekf2.gyro.p * ekf2.gyro_dt/1.e6f;
+  rates_int[1] = ekf2.gyro.q * ekf2.gyro_dt/1.e6f;
+  rates_int[2] = ekf2.gyro.r * ekf2.gyro_dt/1.e6f;
+  accel_int[0] = ekf2.accel.x * ekf2.accel_dt/1.e6f;
+  accel_int[1] = ekf2.accel.y * ekf2.accel_dt/1.e6f;
+  accel_int[2] = ekf2.accel.z * ekf2.accel_dt/1.e6f;
+
+  ekf.setIMUData(stamp, ekf2.gyro_dt, ekf2.accel_dt, rates_int, accel_int);
+
+  if(ekf.attitude_valid()) {
+    // Calculate the quaternion
+    struct FloatQuat ltp_to_body_quat;
+    const Quatf att_q{ekf.calculate_quaternion()};
+    ltp_to_body_quat.qi = att_q(0);
+    ltp_to_body_quat.qx = att_q(1);
+    ltp_to_body_quat.qy = att_q(2);
+    ltp_to_body_quat.qz = att_q(3);
+
+    // Publish it to the state
+    stateSetNedToBodyQuat_f(&ltp_to_body_quat);
+
+    /* Check the quaternion reset state */
+    float delta_q_reset[4];
+    uint8_t quat_reset_counter;
+    ekf.get_quat_reset(delta_q_reset, &quat_reset_counter);
+
+    // If reset update the setpoint heading
+    if(ekf2.quat_reset_counter < quat_reset_counter) {
+      float psi = matrix::Eulerf(matrix::Quatf(delta_q_reset)).psi();
```

Comment:
I still don't get why the filter should have access to the control part. It breaks the separation between estimation and control and make it unusable for a future fixed-wing usage or when using a custom control loop.

Reply:
This will be replaced by an ABI message later on, but for now is really needed for safe operation.

The problem is as following (without this if statement):
- Initial magnetometer measurements are incorrect (for example due to metal in the ground or other interference)
- Therefor the initial heading is incorrect (for example east) and the setpoint is set to this incorrect heading
- At an height of 1.5m the EKF2 filter resets the heading (because it knows initial heading on the ground can be bad)
- Now a correct heading is found (for example north)
- In all control modes, even in attitude you get a sudden yaw movement of the difference between the initial and correct heading. (From north to east) Caused by the incorrect setpoint at boot.

So to solve this correctly there needs to be made a consideration in every control mode if you want to correct the setpoint or not whenever the EKF2 has an quaternion reset. Similar problems occur in hover mode where you want the position setpoint to be updated on a position reset of the EKF2.